src/api/mod.rs

```diff
@@ -97,32 +36,94 @@ where
   Form: Serialize,
 {
   let route = &build_route(path);
-  let client = reqwest::Client::new();
+  let json;
 
-  let json = match type_ {
-    HttpType::Get => {
-      client
-        .get(&build_fetch_query(route, form))
-        .send()
-        .await?
-        .text()
-        .await?
-    }
-    HttpType::Post => client.post(path).json(form).send().await?.text().await?,
-    HttpType::Put => client.put(path).json(form).send().await?.text().await?,
-  };
+  cfg_if! {
+    if #[cfg(feature = "ssr")] {
+     let client = reqwest::Client::new();
+
+     json = match type_ {
+       HttpType::Get => {
+         client
+           .get(&build_fetch_query(route, form))
+           .send()
+           .await?
+           .text()
+           .await?
+       }
+       HttpType::Post => client.post(path).json(form).send().await?.text().await?,
+       HttpType::Put => client.put(path).json(form).send().await?.text().await?,
+     };
+   } else {
+     use wasm_bindgen::UnwrapThrowExt;
+
+     let abort_controller = web_sys::AbortController::new().ok();
+     let abort_signal = abort_controller.as_ref().map(|a| a.signal());
+
+     json = match type_ {
+       HttpType::Get => {
+         gloo_net::http::Request::get(&build_fetch_query(route, form))
+           .abort_signal(abort_signal.as_ref())
+           .send()
+           .await?
+           .text()
+           .await?
+       }
+       HttpType::Post => {
+         gloo_net::http::Request::post(route)
+           .json(form)
+           .expect_throw("Could not parse json body")
+           .abort_signal(abort_signal.as_ref())
+           .send()
+           .await?
+           .text()
+           .await?
+       }
+       HttpType::Put => {
+         gloo_net::http::Request::put(route)
+           .json(form)
+           .expect_throw("Could not parse json body")
+           .abort_signal(abort_signal.as_ref())
+           .send()
+           .await?
+           .text()
+           .await?
+       }
+     };
+
+     // abort in-flight requests if the Scope is disposed
+     // i.e., if we've navigated away from this page
+     leptos::on_cleanup(_cx, move || {
+       if let Some(abort_controller) = abort_controller {
+         abort_controller.abort()
+       }
+     });
+   }
+  }
 
   // Return the error response json as an error
   Response::de(&json).map_err(|_| LemmyAppError::APIError {
     error: json_deser_err(&json),
   })
 }
 
+#[macro_export]
+macro_rules! api_fn {
```

Comment:
I generally hate macros and meta-programming, because you lose compile-time checks, formatting, rust-analyzer support... all to save a few lines of code. Its not necessary here, verbose but checkable is preferred.

Reply:
I removed the macro.